### ISSUE TITLE
shared_embeddings columns BUG: not updated when BP in Estimator which will cause lower metrics in model

### ISSUE THREAD

**Issue by wuxianxingkong** [tensorflow/tensorflow#47037]
When using [share embedding column](https://www.tensorflow.org/api_docs/python/tf/feature_column/shared_embeddings) in premade Estimators, embedding table variables created in it will not be updated in training process.

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Local run or distributed training have the same bug.
- TensorFlow installed from (source or binary):2.4.0
- TensorFlow version (use command below): v2.4.0-rc4-71-g582c8d236cb 2.4.0
- Python version:3.6
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:No cuda. 
- GPU model and memory:No GPU. CPU training.

**Describe the current behavior**

When using tf.feature_column.shared_embeddings in premade Estimator, embedding table variables create in it will not be included in traininng_variables. When I list all variables in checkpoint, Adagrad optimizer will not create any accumulator for shared embedding table variable which means  shared embedding table variable will not  be updated when BP. This bug will cause lower metrics such as AUC compared with old share embedding column in TF1.x with almost the same code.

**Describe the expected behavior**

 shared embedding table variable  should create accumulator in Adagrad or other optmizer.

The reason why this hadppen is that the way shared_embeddings column creates state is not like other feature column such as EmbeddingColumn which using tf.keras.layers.DenseFeatures._state_manager.create_variable, in create_variable function, it will use self._layer.add_weight so that variables can be included in keras.Model.trainable_variables. Finally I have to rewrite shared embedding column code using _state_manager and every thing will be OK.

**Standalone code to reproduce the issue**

```python

import tensorflow as tf
tf.compat.v1.disable_eager_execution()
import pandas as pd
import numpy as np
from tensorflow import feature_column

dataset_url = 'http://storage.googleapis.com/download.tensorflow.org/data/petfinder-mini.zip'
csv_file = 'datasets/petfinder-mini/petfinder-mini.csv'

tf.keras.utils.get_file('petfinder_mini.zip', dataset_url,
                        extract=True, cache_dir='.')
dataframe = pd.read_csv(csv_file)
dataframe['target'] = np.where(dataframe['AdoptionSpeed']==4, 0, 1)

# Drop un-used columns.
dataframe = dataframe.drop(columns=['AdoptionSpeed', 'Description'])

train = dataframe
train_y = dataframe.pop('target')

def input_fn(features, labels, training=True, batch_size=256):
    """An input function for training or evaluating"""
    # Convert the inputs to a Dataset.
    dataset = tf.data.Dataset.from_tensor_slices((dict(features), labels))

    # Shuffle and repeat if you are in training mode.
    if training:
        dataset = dataset.shuffle(1000).repeat()

    return dataset.batch(batch_size)

wide_columns = []
deep_columns = []
# numeric cols
for header in ['PhotoAmt', 'Fee', 'Age']:
  wide_columns.append(feature_column.numeric_column(header))

share_embedding_list = feature_column.shared_embeddings([feature_column.categorical_column_with_hash_bucket('Color1', 10), feature_column.categorical_column_with_hash_bucket('Color2', 10)], dimension=8)

deep_columns.extend(share_embedding_list)

classifier = tf.estimator.DNNLinearCombinedClassifier(
    model_dir='/tmp/model_test',
    linear_feature_columns=wide_columns,
    dnn_feature_columns=deep_columns,
    dnn_hidden_units=[30, 10])
# Train the Model.
classifier.train(
    input_fn=lambda: input_fn(train, train_y, training=True),
    steps=5000)
# Print all variables
save_path ='/tmp/model_test/model.ckpt-5000'
for var in tf.train.list_variables(save_path):
    print(var)

```

**Other info / logs** Include any logs or source code that would be helpful to

Output: 

**('Color1_Color2_shared_embedding', [10, 8])**
('dnn/hiddenlayer_0/bias', [30])
('dnn/hiddenlayer_0/kernel', [16, 30])
('dnn/hiddenlayer_1/bias', [10])
('dnn/hiddenlayer_1/kernel', [30, 10])
('dnn/logits/bias', [1])
('dnn/logits/kernel', [10, 1])
('global_step', [])
('linear/linear_model/Age/weights', [1, 1])
('linear/linear_model/Fee/weights', [1, 1])
('linear/linear_model/PhotoAmt/weights', [1, 1])
('linear/linear_model/bias_weights', [1])
('training/Adagrad/decay', [])
('training/Adagrad/dnn/hiddenlayer_0/bias/accumulator', [30])
('training/Adagrad/dnn/hiddenlayer_0/kernel/accumulator', [16, 30])
('training/Adagrad/dnn/hiddenlayer_1/bias/accumulator', [10])
('training/Adagrad/dnn/hiddenlayer_1/kernel/accumulator', [30, 10])
('training/Adagrad/dnn/logits/bias/accumulator', [1])
('training/Adagrad/dnn/logits/kernel/accumulator', [10, 1])
('training/Adagrad/learning_rate', [])
('training/Ftrl/beta', [])
('training/Ftrl/decay', [])
('training/Ftrl/iter', [])
('training/Ftrl/l1_regularization_strength', [])
('training/Ftrl/l2_regularization_strength', [])
('training/Ftrl/learning_rate', [])
('training/Ftrl/learning_rate_power', [])
('training/Ftrl/linear/linear_model/Age/weights/accumulator', [1, 1])
('training/Ftrl/linear/linear_model/Age/weights/linear', [1, 1])
('training/Ftrl/linear/linear_model/Fee/weights/accumulator', [1, 1])
('training/Ftrl/linear/linear_model/Fee/weights/linear', [1, 1])
('training/Ftrl/linear/linear_model/PhotoAmt/weights/accumulator', [1, 1])
('training/Ftrl/linear/linear_model/PhotoAmt/weights/linear', [1, 1])
('training/Ftrl/linear/linear_model/bias_weights/accumulator', [1])
('training/Ftrl/linear/linear_model/bias_weights/linear', [1])

**Comment by Saduf2019** [comment]
@wuxianxingkong 
I ran your code on tf 2.4 and nightly, please let us now if it confirms your issue.

**Comment by wuxianxingkong** [comment]
Thanks for your reply, yes, it's the same with the output of mine. And I rewrite the code in [notebook](https://colab.research.google.com/gist/wuxianxingkong/0ad5e8867df4738c3734d86707cf3f16/untitled527.ipynb) to compare different behavior between feature_column.embedding_column and feature_column.shared_embeddings. 
``` Python
import tensorflow as tf
tf.compat.v1.disable_eager_execution()
import pandas as pd
import numpy as np
from tensorflow import feature_column

dataset_url = 'http://storage.googleapis.com/download.tensorflow.org/data/petfinder-mini.zip'
csv_file = 'datasets/petfinder-mini/petfinder-mini.csv'

tf.keras.utils.get_file('petfinder_mini.zip', dataset_url,
                        extract=True, cache_dir='.')
dataframe = pd.read_csv(csv_file)
dataframe['target'] = np.where(dataframe['AdoptionSpeed']==4, 0, 1)

# Drop un-used columns.
dataframe = dataframe.drop(columns=['AdoptionSpeed', 'Description'])

train = dataframe
train_y = dataframe.pop('target')

def input_fn(features, labels, training=True, batch_size=256):
    """An input function for training or evaluating"""
    # Convert the inputs to a Dataset.
    dataset = tf.data.Dataset.from_tensor_slices((dict(features), labels))

    # Shuffle and repeat if you are in training mode.
    if training:
        dataset = dataset.shuffle(1000).repeat()

    return dataset.batch(batch_size)

wide_columns = []
deep_columns = []
# numeric cols
for header in ['PhotoAmt', 'Fee', 'Age']:
  wide_columns.append(feature_column.numeric_column(header))
c1 = feature_column.categorical_column_with_hash_bucket('Color1', 10)
c2 = feature_column.categorical_column_with_hash_bucket('Color2', 10)
c1_embdding = feature_column.embedding_column(c1, 4)
share_embedding_list = feature_column.shared_embeddings([c1, c2], dimension=8)

deep_columns.extend(share_embedding_list)
deep_columns.append(c1_embdding)

classifier = tf.estimator.DNNLinearCombinedClassifier(
    model_dir='/tmp/model_test_tf24',
    linear_feature_columns=wide_columns,
    dnn_feature_columns=deep_columns,
    dnn_hidden_units=[30, 10])
# Train the Model.
classifier.train(
    input_fn=lambda: input_fn(train, train_y, training=True),
    steps=5000)
# Print all variables
save_path ='/tmp/model_test_tf24/model.ckpt-5000'
for var in tf.train.list_variables(save_path):
    print(var)


```
 As you can see from the output of [notebook](https://colab.research.google.com/gist/wuxianxingkong/0ad5e8867df4738c3734d86707cf3f16/untitled527.ipynb), I add a c1_embdding column to deep_columns array,  c1_embedding column will create **('dnn/input_from_feature_columns/input_layer/Color1_embedding/embedding_weights', [10, 4])** and **('training/Adagrad/dnn/input_from_feature_columns/input_layer/Color1_embedding/embedding_weights/accumulator', [10, 4])** , but shared_embeddings([c1, c2]) only creates **('Color1_Color2_shared_embedding', [10, 8])**

**Comment by wuxianxingkong** [comment]
@ymodak Any update on this issue?

**Comment by sachinprasadhs** [comment]
Hi, Could you please migrate from Estimator to Keras API which will enable you to perform the aforementioned tasks, such as [model building](https://www.tensorflow.org/guide/keras/custom_layers_and_models), gradient application, [training](https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit), evaluation, and prediction.
Refer Migration guide here https://www.tensorflow.org/guide/migrate/migrating_estimator for more details. Thanks!
